Source/Core/InputCommon/ControllerEmu/StickGate.cpp

```diff
@@ -46,34 +91,145 @@ ControlState SquareStickGate::GetRadiusAtAngle(double ang) const
 {
-  return CalculateInputShapeRadiusAtAngle(ang) * numeric_settings[SETTING_DEADZONE]->GetValue();
+  // FYI: deadzone is scaled by input radius which allows the shape to match.
+  return GetInputRadiusAtAngle(ang) * numeric_settings[SETTING_DEADZONE]->GetValue();
 }
 
 ControlState ReshapableInput::GetInputRadiusAtAngle(double ang) const
 {
-  const ControlState radius =
-      CalculateInputShapeRadiusAtAngle(ang) * numeric_settings[SETTING_INPUT_RADIUS]->GetValue();
-  // Clamp within the -1 to +1 square as input radius may be greater than 1.0:
-  return std::min(radius, SquareStickGate(1).GetRadiusAtAngle(ang));
+  // Handle the "default" state.
+  if (m_calibration.empty())
+  {
+    return GetDefaultInputRadiusAtAngle(ang);
+  }
+
+  const auto sample_pos = ang / MathUtil::TAU * m_calibration.size();
+  // Interpolate the radius between 2 calibration samples.
+  const u32 sample1_index = u32(sample_pos) % m_calibration.size();
+  const u32 sample2_index = (sample1_index + 1) % m_calibration.size();
+  const double sample1_ang = sample1_index * MathUtil::TAU / m_calibration.size();
+  const double sample2_ang = sample2_index * MathUtil::TAU / m_calibration.size();
+
+  return GetRayLineIntersection(
+      GetPointFromAngleAndLength(ang, 1.0),
+      GetPointFromAngleAndLength(sample1_ang, m_calibration[sample1_index]),
+      GetPointFromAngleAndLength(sample2_ang, m_calibration[sample2_index]));
+}
+
+ControlState ReshapableInput::GetDefaultInputRadiusAtAngle(double ang) const
+{
+  // This will normally be the same as the gate radius.
+  // Unless a sub-class is doing weird things with the gate radius (e.g. Tilt)
+  return GetGateRadiusAtAngle(ang);
 }
 
-void ReshapableInput::AddReshapingSettings(ControlState default_radius, ControlState default_shape,
-                                           int max_deadzone)
+void ReshapableInput::SetCalibrationToDefault()
 {
-  // Allow radius greater than 1.0 for definitions of rounded squares
-  // This is ideal for Xbox controllers (and probably others)
-  numeric_settings.emplace_back(
-      std::make_unique<NumericSetting>(_trans("Input Radius"), default_radius, 0, 140));
-  numeric_settings.emplace_back(
-      std::make_unique<NumericSetting>(_trans("Input Shape"), default_shape, 0, 50));
-  numeric_settings.emplace_back(std::make_unique<NumericSetting>(_trans("Dead Zone"), 0, 0, 50));
+  m_calibration.clear();
+}
+
+void ReshapableInput::SetCalibrationToZero()
+{
+  m_calibration.assign(CALIBRATION_SAMPLE_COUNT, 0.0);
+}
+
+void ReshapableInput::SetCalibrationFromGate(const StickGate& gate)
+{
+  m_calibration.resize(gate.GetIdealCalibrationSampleCount().value_or(CALIBRATION_SAMPLE_COUNT));
+
+  u32 i = 0;
+  for (auto& val : m_calibration)
+    val = gate.GetRadiusAtAngle(MathUtil::TAU * i++ / m_calibration.size());
+}
+
+void ReshapableInput::UpdateCalibration(ControlState x, ControlState y)
+{
+  const auto ang_scale = MathUtil::TAU / m_calibration.size();
```

Comment:
Wondering if `ang` vs. `angle` is really worth the save of 2 characters compared to immediate readability?

Reply:
I'll change them all for you. :P